### ISSUE TITLE
Consider an autoenum go generate tool

### ISSUE THREAD

**Issue by jkomoros** [jkomoros/boardgame#467]
Would have behavior like:

```
//go:generate autoenum

//+autoenum
const (
  SuitUnknown = iota
  SuitSpade
  SuitHeart
  SuitClub
  SuitDiamond
)
```
Processes to figure out the longest common prefix in the group and that's the name of the enum. Then the string value is the name minus the prefix, with spaces put between any internal capital letter

Outputs autoenum.go:
```

var Enums = enum.NewSet()

var SuitEnum = Enums.MustAdd("Suit", map[int]string {
  SuitUnknown: "Unknown",
  SuitSpade: "Spade",
  SuitHeart: "Heart",
  SuitClub: "Club",
  SuitDiamond: "SuitDiamond"
})

```

Originally captured in #457

**Comment by jkomoros** [comment]
Maybe consider combining it with autoreader?

**Comment by jkomoros** [comment]
Use https://github.com/abcum/lcp for longest common prefix

**Comment by jkomoros** [comment]
- [ ] Should autoreader name be changed? Slash should it use the general `autoreader:enum` syntax that everybody else uses?
- [ ] Update package doc

**Comment by jkomoros** [comment]
parsing just consts and their doclines shouldn't be hard to do by hand with go.parse

Alternatively, we could have a typed Enum pattern (which will be easier for clients to adopt with the generator):

```
type Color int

const (
  ColorBlue = Color iota
  ColorGreen
  ColorRed
)

var ColorEnum = set.MustAdd("Color", map[int]string{
  int(ColorBlue): "Blue",
  //...
})

//By anonymously embedding we atuomatically implement the MutableEnum interface so the package knows how to deal with them
type TypedColorEnum struct {
  enum.MutableEnum
}

//In your own use you'd only use the Typed Getter/Setter, so the compiler would warn if you used the wrong type of constant
func (t *TypedColorEnum) SetTyped(color Color) error {
  return t.Set(int(color))
}





```

**Comment by jkomoros** [comment]
Actually, you need to embed MutableValues as well as Enums. (Are enums interfaces yet?)

You also need autoreader (and other machinery) to understand that these TypedMutableEnumVal types are actually MutableEnumVal for purposes of Readers.

I can also imagine that this overhead isn't something that everyone will want to do. This Typed pattern seems like a separate issue, that should be supported by AutoReader when it's encountered.

**Comment by jkomoros** [comment]
For our own parser for consts:

Use go/ast package. Filter to only exported. Walk from the top, when you find a GenDecl node that is of type Const and has the magic string in the comment block, iterate through the Specs, cast to *ValueSpec, and read out the Type they are, and the names and Values (unclear if there will be one Spec with multiple names/values or multiple specs each with a name/value.

**Comment by jkomoros** [comment]
Fixed as of 1686f47
